### DISCUSSION TITLE
A directive for BlockedFields

### DISCUSSION THREAD

**Discussion by mjungwir** [graphql-java/graphql-java#2886] [Ideas]
Our project needs a few blocked fields on our schema not exposed to our clients. We implemented using [BlockedFields](https://www.graphql-java.com/documentation/field-visibility/) to success. However, one thing desired was something more declarative in the SDL so one easily understands a field isn't part of the external schema right in the SDL. A directive seems like a nice way to accomplish this desire.

One challenge we faced implementing our custom directive was the `RuntimeWiring` is built, field visibility and directives included, before directives are processed. This prevented us from using `BlockedFields` directly. So we made our own custom `GraphqlFieldVisibility` which holds a Set of field patterns to block. As the directive is processed we add to this Set so when the respective `getFieldDefinitions` or similar on the `GraphqlFieldVisibility` is called it does not return the blocked fields.

Personally I worried this would assume too much about our usage or be incompatible with some of the library. I'm not sure if I'm still concerned looking more into our project and the library itself, but I'm writing this to see if others like the idea and if it's something to pursue. It certainly was compelling to us. Perhaps there's a better or more sustainable way to implement it though?

For extra demonstration, here's how it looks in our project now using an example from the above documentation in the [example project schema](https://github.com/graphql-java/graphql-java-examples/blob/master/http-example/src/main/resources/starWarsSchemaAnnotated.graphqls#L81)

```
# A mechanical creature in the Star Wars universe.
type Droid implements Character {

    appearsIn: [Episode]! @blocked_field

}
```

**Comment by bbakerman** [comment]
Have  a look at `graphql.schema.SchemaTransformer` and its usage in `graphql.schema.transform.FieldVisibilitySchemaTransformation`.

This allows you to take a schema and "transform" it so fields and their referenced types are completely removed while fixing up usages etc... (eg a type will be removed if all fields that reference it are removed etc...)

This is better than `GraphqlFieldVisibility` because its static - no runtime penalty - no introspection leakage etc..

You would create a predicate that looked for your special `@blocked_field` and then removed the field

**Reply by mjungwir** [reply]
Thanks @bbakerman, that worked well! That would be nice to document on the [Field Visibiilty](https://www.graphql-java.com/documentation/field-visibility) page IMO. 

What are your or others thoughts on a library `@blockeField` directive "out of the box" to supply this feature? It's certainly not hard to implement with this `FieldVisibilitySchemaTransformation` so perhaps that's the preference as custom directives give more ability to specify why it's blocked, etc.?

**Reply by bbakerman** [reply]
I dont think we would put in a specific directive out of the box.  We have the tools via the FieldVisibilitySchemaTransformation and it would be a bike shedding exercise to try and come up with a canonical semantic directive.
